pkg/chains/storage/grafeas/grafeas.go

```diff
@@ -0,0 +1,354 @@
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	envelope := &commonpb.Envelope{
+		Payload:     payload,
+		PayloadType: "in-toto attestations containing a slsa.dev/provenance predicate",
+		Signatures: []*commonpb.EnvelopeSignature{
+			{
+				Sig:   []byte(signature),
+				Keyid: b.cfg.Signers.KMS.KMSRef, // TODO: currently we only support storing kms keyID, will add other keys' ids later i.e. k8s secret, fulcio
+			},
+		},
+	}
+
+	occurrence := &pb.Occurrence{
+		Resource: &pb.Resource{
+			// namespace-scoped resource
+			// https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-uris
+			Uri: b.retrieveResourceURI(opts),
+		},
+		NoteName: b.getNotePath(),
+		Details:  occurrenceDetails,
+		Envelope: envelope,
+	}
+
+	occurrenceRequest := &pb.CreateOccurrenceRequest{
+		Parent:     b.getProjectPath(),
+		Occurrence: occurrence,
+	}
+
+	return occurrenceRequest
+}
+
+func (b *Backend) getProjectPath() string {
+	projectID := b.cfg.Storage.Grafeas.ProjectID
+	return fmt.Sprintf(ProjectNameFormat, projectID)
+}
+
+func (b *Backend) getNotePath() string {
+	projectID := b.cfg.Storage.Grafeas.ProjectID
+	noteID := b.cfg.Storage.Grafeas.NoteID
+	return fmt.Sprintf(NoteNameFormat, projectID, noteID)
+}
+
+// decide the attestation content type based on its format (simplesigning or in-toto)
+func (b *Backend) getContentType(opts config.StorageOpts) attestationpb.GenericSignedAttestation_ContentType {
+	// for simplesigning
+	if opts.PayloadFormat == formats.PayloadTypeSimpleSigning {
+		return attestationpb.GenericSignedAttestation_SIMPLE_SIGNING_JSON
+	}
+
+	// for in-toto
+	return attestationpb.GenericSignedAttestation_CONTENT_TYPE_UNSPECIFIED
+}
+
+// retrieve all occurrences using the list of auto-generated occurrence names that were stored previously
+func (b *Backend) getOccurrences(ctx context.Context) ([]*pb.Occurrence, error) {
+	result := []*pb.Occurrence{}
+
+	for _, occName := range b.occurrenceRefs {
+		getOCCReq := &pb.GetOccurrenceRequest{
+			Name: occName,
+		}
+		occ, error := b.client.GetOccurrence(ctx, getOCCReq)
+		if error != nil {
+			return nil, error
+		}
+		result = append(result, occ)
+	}
+
+	return result, nil
+}
+
+// addTypeInformationToObject adds TypeMeta information to a runtime.Object based upon the loaded scheme.Scheme
+// inspired by: https://github.com/kubernetes/cli-runtime/blob/v0.19.2/pkg/printers/typesetter.go#L41
+// Souce: https://github.com/kubernetes/client-go/issues/308#issuecomment-700099260
+func addTypeInformationToObject(obj runtime.Object) error {
+	gvks, _, err := scheme.Scheme.ObjectKinds(obj)
+	if err != nil {
+		return fmt.Errorf("missing apiVersion or kind and cannot assign it; %w", err)
+	}
+
+	for _, gvk := range gvks {
+		if len(gvk.Kind) == 0 {
+			continue
+		}
+		if len(gvk.Version) == 0 || gvk.Version == runtime.APIVersionInternal {
+			continue
+		}
+		obj.GetObjectKind().SetGroupVersionKind(gvk)
```

Comment:
Can we return the GVK instead of modifying the incoming Object? I'm a bit nervous about modifying the original content that we're using to sign.

Reply:
Yes, I just found the correct method to get gvk. here is the commit a0c2680ac800e9c02c61ee5e1c287321e3e93363